Move GenTreeIntCon::gtCompileTimeHandle field to be a global map

EgorBo opened the pull request:
This PR removes `GenTreeIntCon::gtCompileTimeHandle` field in favor of a global hashmap (we already had it in VN).

- This allows to remove a TODO where we used to give on 0 compile time handles in VN
- Fixes a bug @kunalspathak found

Fixes: https://github.com/dotnet/runtime/issues/114823

jakobbotsch commented:
> we already had it in VN

Having internal maps in phases that keep them maintained is not as problematic as having global ambient state throughout the JIT that must be maintained by all phases and becomes part of the IR invariants.

Though, given that this is a map from the value itself and not from `GenTree*`, it probably does not come with the same downsides, so perhaps it's not too bad.

SingleAccretion commented:
An alternative to this would be to introduce a JIT-EE API for `embedded handle -> compile time handle`.

EgorBo committed:
remove unused JIT-EE APIs

EgorBo committed:
clean up

EgorBo committed:
Fix the issue

EgorBo committed:
Update jiteeversionguid.h

EgorBo committed:
Update jiteeversionguid.h

EgorBo commented:
Ok I ended up reverting my attempts to either use a global hash or JIT-EE - they both were not trivial, and just removed unused JIT-EE calls and add "!=0" checks to the VN's compilteTimeHandle cache. We already had that `!=0` check here 
https://github.com/dotnet/runtime/blob/dcf79a484f0c19ea70e58a38cee2e9d5e1939cec/src/coreclr/jit/valuenum.cpp#L4696-L4706

I just added it to other uses of that map.


PTAL @jakobbotsch @kunalspathak @dotnet/jit-contrib

jakobbotsch commented on src/coreclr/inc/corinfo.h:
```diff
@@ -1204,7 +1204,7 @@ enum CorInfoAccessAllowedHelperArgType
     CORINFO_HELPER_ARG_TYPE_Method  = 2,
     CORINFO_HELPER_ARG_TYPE_Class   = 3,
-    CORINFO_HELPER_ARG_TYPE_Module  = 4,
+    // CORINFO_HELPER_ARG_TYPE_Module  = 4,
```
Can we remove it and change the values if we are changing JIT-EE GUID anyway?

EgorBo replied:
Doesn't it need r2r bump then?

EgorBo committed:
Address feedback

EgorBo committed:
Merge branch 'main' into remove-compile-time-handle-fld

EgorBo committed:
Merge branch 'main' into remove-compile-time-handle-fld

jkotas commented on src/coreclr/jit/compiler.hpp:
```diff
@@ -1489,13 +1489,7 @@ inline GenTreeIntCon* Compiler::gtNewIconHandleNode(size_t value, GenTreeFlags f
-    assert((!embedScpHnd) != (!pEmbedScpHnd));
-
-    return gtNewIconEmbHndNode(embedScpHnd, pEmbedScpHnd, GTF_ICON_SCOPE_HDL, scpHnd);
+    return gtNewIconEmbHndNode((void*)scpHnd, nullptr, GTF_ICON_SCOPE_HDL, scpHnd);
```
Isn't this introducing an abstraction violation? `CORINFO_XXX_HANDLE`s are not meant to be embedded in the code. We have an intentional distinction between the `CORINFO_XXX_HANDLE` handles used to represent type system entities on JIT/EE interface and how these values get embedded into the code.

It happens to work since these module handles are only used with JIT currently, but this change does not look like an improvement to me.

EgorBo replied:
@jkotas I just assumed that almost 300 LOC is too much for an API that is just no-op (and CoreCLR) only, I can revert this part.

Looks like the only case where we need to embed a field handle is CoreCLR jitting a method that accesses a field it has no access for, so JIT emits an exception telling about the field access violation. Isn't part of the IL verification that we got rid some time ago?

jkotas commented on src/coreclr/jit/compiler.hpp:
```diff
@@ -1528,13 +1529,7 @@ inline GenTree* Compiler::gtNewIconEmbMethHndNode(CORINFO_METHOD_HANDLE methHnd)
-    assert((!embedFldHnd) != (!pEmbedFldHnd));
-
-    return gtNewIconEmbHndNode(embedFldHnd, pEmbedFldHnd, GTF_ICON_FIELD_HDL, fldHnd);
+    return gtNewIconEmbHndNode((void*)fldHnd, nullptr, GTF_ICON_FIELD_HDL, fldHnd);
```
Same here
